### PR TITLE
server/schedule: refine metrics about store limit

### PR DESCRIPTION
**Pull request by wangrzneu**
### What problem does this PR solve? 
refine metrics about store limit: add storeLimitCostCounter and storeLimitAvailableGauge to monitor the status of store limiter.

### What is changed and how it works?
refine metrics about store limit
1. add storeLimitCostCounter to monitor the step cost
2. add storeLimitAvailableGauge to monitor the available status of storeLimit, and it is collected in a background job.
3. add storeLimitRateGauge to show the store limit setting in pd.

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)
1. run a tidb cluste with 4 tikv instances
2. prepare some data using sysbench like : `sysbench --mysql-host=127.0.0.1 --mysql-port=4000 --mysql-user=root --table-size=1000000 oltp_insert prepare`
3. run `pd-ctl -u 0.0.0.0:32774 store limit all 1  region-add`  and `pd-ctl -u 0.0.0.0:32774 store limit all 1  region-remove ` to set the store limit
4. run `pd-ctl -u 0.0.0.0:32774 store delete 3057` to remove one store
5. run `curl 0.0.0.0:32774/metrics | grep store_limit` to get the metrics. The metrics should be like:
```
# HELP pd_schedule_store_limit_available available limit rate of store.
# TYPE pd_schedule_store_limit_available gauge
pd_schedule_store_limit_available{limit_type="region-add",store="1"} 1
pd_schedule_store_limit_available{limit_type="region-add",store="3057"} 1
pd_schedule_store_limit_available{limit_type="region-add",store="4"} 1
pd_schedule_store_limit_available{limit_type="region-add",store="5"} 1
pd_schedule_store_limit_available{limit_type="region-remove",store="1"} 1
pd_schedule_store_limit_available{limit_type="region-remove",store="3057"} 100000
pd_schedule_store_limit_available{limit_type="region-remove",store="4"} 1
pd_schedule_store_limit_available{limit_type="region-remove",store="5"} 1
# HELP pd_schedule_store_limit_cost limit rate cost of store.
# TYPE pd_schedule_store_limit_cost counter
pd_schedule_store_limit_cost{limit_type="region-add",store="1"} 1
pd_schedule_store_limit_cost{limit_type="region-add",store="4"} 2
pd_schedule_store_limit_cost{limit_type="region-remove",store="3057"} 3
# HELP pd_schedule_store_limit_rate the limit rate of store.
# TYPE pd_schedule_store_limit_rate gauge
pd_schedule_store_limit_rate{limit_type="region-add",store="1"} 1.000000016666667
pd_schedule_store_limit_rate{limit_type="region-add",store="3057"} 6e+06
pd_schedule_store_limit_rate{limit_type="region-add",store="4"} 1.000000016666667
pd_schedule_store_limit_rate{limit_type="region-add",store="5"} 1.000000016666667
pd_schedule_store_limit_rate{limit_type="region-remove",store="1"} 1.000000016666667
pd_schedule_store_limit_rate{limit_type="region-remove",store="3057"} 6e+06
pd_schedule_store_limit_rate{limit_type="region-remove",store="4"} 1.000000016666667
pd_schedule_store_limit_rate{limit_type="region-remove",store="5"} 1.000000016666667
```
 
Side effects

 - Breaking backward compatibility
The metrics pd_schedule_store_limit is removed.

### CODE REVIEW

**Comment by rleungx** [comment]
I am not sure the metrics of cost or available token is useful (due to the time interval of metrics collection?), but we can record the capacity of the store limit to avoid using pd-ctl to find it.

**Comment by wangrzneu** [comment]
> I am not sure the metrics of cost or available token is useful (due to the time interval of metrics collection?), but we can record the capacity of the store limit to avoid using pd-ctl to find it. What do you think?

I add a metric pd_schedule_store_limit_rate to show the store limit setting of pd.

File: server/schedule/operator_controller.go [hunk 1]
```diff
@@ -973,3 +972,30 @@ func (oc *OperatorController) RemoveStoreLimit(storeID uint64) {
 	}
 	delete(oc.storesLimit, storeID)
 }
+
+// CollectStoreLimitMetrics collects the metrics about store limit
+func (oc *OperatorController) CollectStoreLimitMetrics() {
```
Comment (shafreeck) [inline server/schedule/operator_controller.go:977]: Where is the new added function called?

Comment (wangrzneu) [reply server/schedule/operator_controller.go:977]: > Where is the new added function called?

It is called in server/cluster/cluster.go:115.

**Comment by nolouch** [comment]
PTAL @rleungx

File: server/schedule/metrics.go [hunk 2]
```diff
@@ -50,19 +50,37 @@ var (
...
 		}, []string{"type"})
 
-	storeLimitGauge = prometheus.NewGaugeVec(
+	storeLimitAvailableGauge = prometheus.NewGaugeVec(
 		prometheus.GaugeOpts{
 			Namespace: "pd",
 			Subsystem: "schedule",
-			Name:      "store_limit",
-			Help:      "Limit of store.",
-		}, []string{"store", "type", "limit_type"})
```
Comment (nolouch) [inline server/schedule/metrics.go:59]: The metrics will not be compatible with the old version of grafana template, why can't it keep the original way?

File: server/schedule/operator_controller.go [hunk 3]
```diff
@@ -973,3 +972,30 @@ func (oc *OperatorController) RemoveStoreLimit(storeID uint64) {
...
+				continue
+			}
+			for n, v := range storelimit.TypeNameValue {
+				storeLimit := oc.storesLimit[storeID][v]
+				if storeLimit == nil {
+					continue
+				}
+				storeIDStr := strconv.FormatUint(storeID, 10)
+				storeLimitAvailableGauge.WithLabelValues(storeIDStr, n).Set(float64(storeLimit.Available()) / float64(storelimit.RegionInfluence[v]))
+				storeLimitRateGauge.WithLabelValues(storeIDStr, n).Set(storeLimit.Rate() * StoreBalanceBaseTime)
```
Comment (nolouch) [inline server/schedule/operator_controller.go:997]: Users may more likely want to know the store how many operators can be produced in one minute, for example, default is 15 opm.

Comment (wangrzneu) [reply server/schedule/operator_controller.go:997]: The metrics `storeLimitRateGauge` can show the operator producing rate.

File: server/schedule/metrics.go (hunk 2 above)
Comment (wangrzneu) [reply server/schedule/metrics.go:59]: > The metrics will not be compatible with the old version of grafana template, why can't it keep the original way?

The old way has two `type`, `"take"` and `"available"`. I use  a new metrics `storeLimitAvailableGauge` to record the `"available"` type `storeLimitGauge`, but use a different metrics `storeLimitCostCounter` (it is a counter, not gauge) to record the original `"available"` type `storeLimitGauge` metrics.

File: server/schedule/operator_controller.go (hunk 3 above)
Comment (nolouch) [reply server/schedule/operator_controller.go:997]: got.

**Pull request merged**